File: 3.x.x/crates-utopia/emails/src/validator/email_not_disposable.rs
use serde_json::Value;
use utopia_validators::{Validator, ValueType};

use crate::Email;

/// Validate that an email is not disposable (PHP `EmailNotDisposable`).
#[derive(Debug, Clone, Default)]
pub struct EmailNotDisposable;

impl EmailNotDisposable {
    /// Create a non-disposable validator.
    pub fn new() -> Self {
        Self
    }
}

impl Validator for EmailNotDisposable {
    fn description(&self) -> String {
        "Value must be a valid email address that is not from a disposable email service".into()
    }

    fn is_array(&self) -> bool {
        false
    }

    fn value_type(&self) -> ValueType {
        ValueType::String
    }

    fn is_valid(&self, value: &Value) -> bool {
        let Some(raw) = value.as_str() else {
            return false;
        };
        Email::new(raw)
            .map(|email| email.is_valid() && !email.is_disposable())
            .unwrap_or(false)
    }
}

File: packages/emails/src/Validator/EmailNotDisposable.php
<?php

namespace Utopia\Emails\Validator;

use Utopia\Emails\Email;
use Utopia\Validator;

/**
 * EmailNotDisposable
 *
 * Validate that an email address is not from a disposable email service
 */
class EmailNotDisposable extends Validator
{
    /**
     * Get Description
     *
     * Returns validator description
     */
    public function getDescription(): string
    {
        return 'Value must be a valid email address that is not from a disposable email service';
    }

    /**
     * Is valid
     *
     * Validation will pass when $value is a valid email address that is not disposable
     *
     * @param  mixed  $value
     */
    public function isValid($value): bool
    {
        if (! is_string($value)) {
            return false;
        }

        try {
            $email = new Email($value);

            return $email->isValid() && ! $email->isDisposable();
        } catch (\Exception $e) {
            return false;
        }
    }

    /**
     * Is array
     *
     * Function will return true if object is array.
     */
    public function isArray(): bool
    {
        return false;
    }

    /**
     * Get Type
     *
     * Returns validator type.
     */
    public function getType(): string
    {
        return self::TYPE_STRING;
    }
}
